internal/sql/rewrite/parameters.go

```diff
@@ -41,55 +41,61 @@ func isNamedParamSignCast(node ast.Node) bool {
 	return astutils.Join(expr.Name, ".") == "@" && cast
 }
 
-func NamedParameters(engine config.Engine, raw *ast.RawStmt, numbs map[int]bool, dollar bool) (*ast.RawStmt, map[int]string, []source.Edit) {
+// paramFromName takes a user-defined parameter name, with an optional suffix of
+// ? (nullable), or ! (non-null) and builds the appropiate parameter
+func paramFromName(name string) named.Param {
+	if len(name) == 0 {
+		return named.NewUnspecifiedParam(name)
+	}
+
+	last := name[len(name)-1]
+	if last == '!' {
+		return named.NewUserDefinedParam(name[:len(name)-1], true)
+	}
+
+	if last == '?' {
+		return named.NewUserDefinedParam(name[:len(name)-1], false)
+	}
+
+	return named.NewUnspecifiedParam(name)
```

Comment:
This is the core logic for the nullability options, I'm happy to rip this out and replace it with anything else.